gc/base/MemoryManager.cpp

```diff
@@ -244,8 +250,35 @@ MM_MemoryManager::createVirtualMemoryForHeap(MM_EnvironmentBase *env, MM_MemoryH
 
 						if ((NULL == instance) && (ceiling > (void *)NON_SCALING_LOW_MEMORY_HEAP_CEILING)) {
 
+#if defined(S390)
+							/*
+							 * Z platform executes smaller shift faster, so it is beneficial to have smaller one.
+							 * Attempting to allocate object heap below 8G or 16G
+							 * to get Compressed References shift 1 or 2 for zLinux only.
+							 */
+
+#define EIGHT_GB_ADDRESS ((void *)((uintptr_t)8 * 1024 * 1024 * 1024))
+#define SIXTEEN_GB_ADDRESS ((void *)((uintptr_t)16 * 1024 * 1024 * 1024))
+
+							/*
+							 * Attempt to allocate heap below 8G.
+							 */
+							if ((NULL == instance) && (requestedTopAddress <= EIGHT_GB_ADDRESS)) {
+								instance = MM_VirtualMemory::newInstance(env, heapAlignment, allocateSize, pageSize, pageFlags, tailPadding, preferredAddress,
+										(void *)OMR_MIN((uintptr_t)EIGHT_GB_ADDRESS, (uintptr_t)ceiling), mode, options, memoryCategory);
```

Comment:
to make it easier to read, ceiling could be recalculated at a separate line with a new name (for example adjustedCeiling)